### PR TITLE
Support Partition Table Writing When bypass TiDB

### PR DESCRIPTION
**Pull request by Daemonxiao**
## What is the purpose of the change

This pr is a backport feature from TiSpark about [bypass-TiDB writing partition table](https://github.com/pingcap/tispark/issues/2437).

**TiSpark support bypass-TiDB writing partition table list**
- [Design](https://github.com/pingcap/tispark/blob/master/docs/design/2022-06-30-partition-table-batchwrite-suppport.md)
- [Code](https://github.com/pingcap/tispark/pull/2451)
- [Doc](https://github.com/pingcap/tispark/blob/master/docs/userguide.md#partition-table-support)

## Brief change log

- Implementation of how to calculate partition table id.
- When the table is a partition table, use the partition table id as the physical id in the process of calculating key.

## Verifying this change

This change added tests and can be verified as follows:

- Partition table writing test
- Partition table deleting test

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): (yes) Just add the Lombok dependence with the `provided` scope.
- The public API: (no)
- The runtime per-record code paths (performance sensitive): (yes)

## Documentation

- Does this pull request introduces a new feature? (yes)
- If yes, how is the feature documented? (docs) It is in `docs/flink_partition_table.md`.

### CODE REVIEW

File: tidb/src/main/java/io/tidb/bigdata/tidb/partition/PartitionedTable.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+              tiPartitionDef.getId(),
+              logicalTable.getTableInfo());
+    }
+
+    return new PartitionedTable(
+        logicalTable, physicalTables, partitionExpr, tableInfo.getPartitionInfo());
+  }
+
+  private static long getLogicalTableId() {
+    return 0;
```
Comment (xuanyu66) [inline tidb/src/main/java/io/tidb/bigdata/tidb/partition/PartitionedTable.java:77]: why return 0?

File: flink/README.md [hunk 2]
```diff
@@ -150,13 +150,16 @@ Received a total of 1 row
...
+## Partition table supports
+
+See [flink_partition_table](../docs/flink_partition_table.md)
+
 ## Supports and Limitations
 
 TiBigDate/Flink supports writing data to and reading from clustered index tables, which is a new feature in TiDB-5.0.0.
 
 TiBigDate/Flink does not support the following features: 
 - bypass-write tables with auto random column
```
Comment (xuanyu66) [inline flink/README.md:162]: Maybe delete this too
https://github.com/tidb-incubator/TiBigData/pull/234

**Review by xuanyu66 — CHANGES_REQUESTED** [review]

File: tidb/src/main/java/io/tidb/bigdata/tidb/partition/PartitionedTable.java (hunk 1 above)
Comment (Daemonxiao) [reply tidb/src/main/java/io/tidb/bigdata/tidb/partition/PartitionedTable.java:77]: @xuanyu66 Thanks for your review. It's a useless function, and I will delete it.

File: flink/README.md (hunk 2 above)
Comment (Daemonxiao) [reply flink/README.md]: OK, I have deleted this.

**Review by shiyuhang0 — APPROVED** [review]

**Review by xuanyu66 — APPROVED** [review]

File: tidb/src/main/java/io/tidb/bigdata/tidb/partition/PartitionedTable.java [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+      // Hash partition only support int types, they all can convert to Number
+      Number id = (Number) row.get(columnRef.getColumnOffset(), columnRef.getDataType());
+      int partitionId = (int) (id.longValue() % physicalTables.length);
+      return physicalTables[partitionId];
+    } else if (originalExpr instanceof FuncCallExpr) {
+      // TODO: support more function partition
+      FuncCallExpr partitionFuncExpr = (FuncCallExpr) originalExpr;
+      if (partitionFuncExpr.getFuncTp() == YEAR) {
+        int result =
+            (int) partitionFuncExpr.eval(Constant.create(row.getDate(0), DateType.DATE)).getValue();
```
Comment (xuanyu66) [inline tidb/src/main/java/io/tidb/bigdata/tidb/partition/PartitionedTable.java:177]: fix this bug

Comment (Daemonxiao) [reply tidb/src/main/java/io/tidb/bigdata/tidb/partition/PartitionedTable.java]: done

**Pull request merged**